### PR TITLE
Authentication Implementation

### PR DESCRIPTION
**Pull request by Raisondetr3**
## Description

This pull request introduces the full implementation of the authentication functionality into the `dev` branch. The changes include user registration, login, and integration of JWT-based authentication. Additionally, comprehensive unit and integration tests are added to ensure functionality and reliability. Swagger documentation has been provided for API endpoints.

---

### Authentication API
- **User Registration**:
  - Endpoint: `POST /api/v1/auth/register`
  - Allows new users to register by providing a username, email, and password.
  - Validates the uniqueness of usernames and ensures password security using BCrypt encoding.

- **User Login**:
  - Endpoint: `POST /api/v1/auth/login`
  - Authenticates users by verifying credentials and returns a JWT token with an expiration time.

### Security
- Implemented **JWT-based authentication** for stateless security.
- Configured `SecurityConfig` for:
  - Stateless session management.
  - JWT token validation and user authentication.
- **Role-based access control** with endpoints accessible for authenticated users only.

### Tests
- **Unit Tests**:
  - Coverage for all service methods, including edge cases.
- **Integration Tests**:
  - Comprehensive tests for `UserController` to validate endpoint functionality and error handling.

### Documentation
- Added Swagger documentation for the API.
- Accessible at: `/swagger-ui.html`.
- Detailed descriptions of API endpoints, request/response structures, and example payloads.
- Added javadoc documentation for classes.

### New Files
1. `UserController.java`: Handles API requests for registration and login.
2. `JwtService.java`: Utility for generating and validating JWT tokens.
3. `UserService.java`: Handles business logic for user authentication.
4. `OpenApiConfig.java`: Configures OpenAPI/Swagger documentation.

### Tests
1. `UserServiceTest.java`: Unit tests for user service methods.
2. `UserControllerIntegrationTest.java`: Integration tests for the authentication API.

### Configurations
1. Created `SecurityConfig`:
   - Configures security filters using `SecurityFilterChain` to enforce JWT-based authentication and stateless session management.
   - Defines `JwtAuthenticationFilter` for processing JWT tokens.
   - Includes beans for `AuthenticationProvider`, `AuthenticationManager`, and `PasswordEncoder` to support secure authentication and password encoding.
   - Adds a custom `UserDetailsService` implementation for user lookup from the database.
2. Updated `application-test.yml`:
   - JWT secret key and expiration time properties.

---

## How to Test
1. Run unit tests:
   ```bash
   mvn test
   ```

2. Run integration tests:
   ```bash
   mvn verify
   ```

3. Start the application:
   ```bash
   mvn spring-boot:run
   ```

3. Use Swagger UI to test endpoints::
   ```bash
   Navigate to: http://localhost:8080/swagger-ui.html
   ```

## Checklist

- [x] Implemented user registration and login.
- [x] Integrated JWT authentication.
- [x] Added Swagger documentation.
- [x] Created unit and integration tests.

### CODE REVIEW

File: google_checks.xml [hunk 1]
```diff
@@ -54,7 +54,7 @@
 
   <module name="LineLength">
     <property name="fileExtensions" value="java"/>
-    <property name="max" value="100"/>
+    <property name="max" value="115"/>
```
Comment (XVIIStarPlatinum) [inline google_checks.xml:57]: https://www.google.com/url?sa=i&url=https%3A%2F%2Fwww.mirror.co.uk%2Fsport%2Ffootball%2Fnews%2Fman-city-115-charges-hearing-33449283&psig=AOvVaw1Y92bQW5mu1Hc7pA0UG2WJ&ust=1733259576132000&source=images&cd=vfe&opi=89978449&ved=0CBcQjhxqFwoTCNCW2pz9iYoDFQAAAAAdAAAAABAn

**Review by XVIIStarPlatinum — APPROVED** [review]
A little bit of trolling with the tab spaces (+115 reference). But otherwise, LGTM
